tao_compiler/mlir/disc/transforms/disc_specialize_fusion_with_speculation.cc

```diff
@@ -461,48 +461,66 @@ struct DiscSpecializeFusionWithSpeculationPass
                                : 1;
       auto max_rows_per_wave =
           max_threads_per_wave / block_size * row_per_block;
-      threshold = b.create<arith::ConstantIndexOp>(loc, max_rows_per_wave);
+      Value threshold =
+          b.create<arith::ConstantIndexOp>(loc, max_rows_per_wave);
       Value operand = dominant_equivalent_op->getOperand(0);
       // #out-element is row numbers.
-      out_element_number = b.create<memref::DimOp>(loc, operand, 0);
+      Value out_element_number = b.create<memref::DimOp>(loc, operand, 0);
+
+      Value larger = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::sgt,
+                                             out_element_number, threshold);
+      Value divisible = b.create<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::eq,
+          b.create<arith::RemUIOp>(
+              loc, out_element_number,
+              b.create<arith::ConstantIndexOp>(loc, vector_size)),
+          b.create<arith::ConstantIndexOp>(loc, 0));
+      pred = b.create<arith::AndIOp>(loc, larger, divisible);
     } else if (fusion_type == FusionType::kLoop) {
-      threshold = b.create<arith::ConstantIndexOp>(loc, max_threads_per_wave);
       FusionPatternBase fusion_pattern(fusion_op);
       Operation* dominant_equivalent_op = fusion_pattern.getRootOps().back();
-      out_element_number =
+      Value out_element_number =
           emitNumElementsComputation(b, loc, dominant_equivalent_op);
+      if (mem_intensive_opt_experimental) {
+        vector_size = 4;
```

Comment:
Maybe the vector_size should depend on the element type, e.g. vector_size = 128 / element_bit_width ?

Reply:
The vector-size also affects the thread level parallelism. I will check the performance of different sizes when vector-size is 8 for FP16.

Reply:
OK, this is a tradeoff. I think it's OK to leave the vector-size decision for fp16 as a TODO for now.
AFAIK there's not any benifit to have a vector width larger than 128 bit, e.g. vector_szie=4 for fp64. What do you think?

Reply:
Updated. Current commit set the vector-size according to 128 / min-bit-width-of-outputs. I have check the performance. It is good for a small shape like `128 x 1024`. The case of tiny shape hurts (e.g., `8 x 1024`), but the penalty is less than a us, which I think is OK.

Reply:
Great!